File: insuranceApp/backend/app/schemas/user_info.py
import uuid
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class BasicInfo(BaseModel):
    """用户基本信息"""
    age: Optional[str] = Field(None, description="年龄")
    city: Optional[str] = Field(None, description="城市")
    gender: Optional[str] = Field(None, description="性别")


class FinancialInfo(BaseModel):
    """用户财务信息"""
    occupation: Optional[str] = Field(None, description="职业")
    income: Optional[str] = Field(None, description="收入")
    expenses: Optional[str] = Field(None, description="支出")
    assets: Optional[str] = Field(None, description="资产")
    liabilities: Optional[str] = Field(None, description="负债")


class RiskInfo(BaseModel):
    """用户风险信息"""
    risk_aversion: Optional[str] = Field(None, description="风险厌恶程度")


class RetirementInfo(BaseModel):
    """用户退休信息"""
    retirement_age: Optional[str] = Field(None, description="退休年龄")
    retirement_income: Optional[str] = Field(None, description="退休收入")


class FamilyMember(BaseModel):
    """家庭成员"""
    relation: Optional[str] = Field(None, description="关系")
    age: Optional[str] = Field(None, description="年龄")
    occupation: Optional[str] = Field(None, description="职业")
    income: Optional[str] = Field(None, description="收入")


class FamilyInfo(BaseModel):
    """用户家庭信息"""
    family_members: Optional[List[FamilyMember]] = Field(default_factory=list, description="家庭成员列表")


class Goal(BaseModel):
    """目标"""
    goal_details: Optional[str] = Field(None, description="目标详情")


class GoalInfo(BaseModel):
    """用户目标信息"""
    goals: Optional[List[Goal]] = Field(default_factory=list, description="目标列表")


class OtherInfo(BaseModel):
    """其他信息"""
    pass


class UserInfoBase(BaseModel):
    """用户个人信息基础模型"""
    basic_info: Optional[BasicInfo] = Field(default_factory=BasicInfo, description="基本信息")
    financial_info: Optional[FinancialInfo] = Field(default_factory=FinancialInfo, description="财务信息")
    risk_info: Optional[RiskInfo] = Field(default_factory=RiskInfo, description="风险信息")
    retirement_info: Optional[RetirementInfo] = Field(default_factory=RetirementInfo, description="退休信息")
    family_info: Optional[FamilyInfo] = Field(default_factory=FamilyInfo, description="家庭信息")
    goal_info: Optional[GoalInfo] = Field(default_factory=GoalInfo, description="目标信息")
    other_info: Optional[Dict[str, Any]] = Field(default_factory=dict, description="其他信息")


class UserInfoGetRequest(BaseModel):
    """获取用户个人信息请求"""
    user_id: uuid.UUID = Field(..., description="用户ID")


class UserInfoCreate(UserInfoBase):
    """创建用户个人信息"""
    user_id: uuid.UUID = Field(..., description="用户ID")


class UserInfoUpdate(UserInfoBase):
    """更新用户个人信息"""
    user_id: uuid.UUID = Field(..., description="用户ID")


class UserInfoInDB(UserInfoBase):
    """数据库中的用户个人信息"""
    user_id: uuid.UUID = Field(..., description="用户ID")
    
    class Config:
        from_attributes = True


class UserInfoResponse(UserInfoInDB):
    """用户个人信息响应"""
    pass


class ApiResponse(BaseModel):
    """API响应基础模型"""
    code: int = Field(..., description="状态码")
    message: str = Field(..., description="响应消息")


class UserInfoApiResponse(ApiResponse):
    """用户个人信息API响应"""
    user_info: Optional[UserInfoResponse] = Field(None, description="用户个人信息") 

File: insuranceApp/frontend/lib/models/user_info.dart
import 'dart:convert';

class BasicInfo {
  String? age;
  String? city;
  String? gender;

  BasicInfo({
    this.age,
    this.city,
    this.gender,
  });

  factory BasicInfo.fromJson(Map<String, dynamic> json) {
    return BasicInfo(
      age: json['age'],
      city: json['city'],
      gender: json['gender'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'age': age,
      'city': city,
      'gender': gender,
    };
  }

  BasicInfo copyWith({
    String? age,
    String? city,
    String? gender,
  }) {
    return BasicInfo(
      age: age ?? this.age,
      city: city ?? this.city,
      gender: gender ?? this.gender,
    );
  }
}

class FinancialInfo {
  String? occupation;
  String? income;
  String? expenses;
  String? assets;
  String? liabilities;

  FinancialInfo({
    this.occupation,
    this.income,
    this.expenses,
    this.assets,
    this.liabilities,
  });

  factory FinancialInfo.fromJson(Map<String, dynamic> json) {
    return FinancialInfo(
      occupation: json['occupation'],
      income: json['income'],
      expenses: json['expenses'],
      assets: json['assets'],
      liabilities: json['liabilities'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'occupation': occupation,
      'income': income,
      'expenses': expenses,
      'assets': assets,
      'liabilities': liabilities,
    };
  }

  FinancialInfo copyWith({
    String? occupation,
    String? income,
    String? expenses,
    String? assets,
    String? liabilities,
  }) {
    return FinancialInfo(
      occupation: occupation ?? this.occupation,
      income: income ?? this.income,
      expenses: expenses ?? this.expenses,
      assets: assets ?? this.assets,
      liabilities: liabilities ?? this.liabilities,
    );
  }
}

class RiskInfo {
  String? riskAversion;

  RiskInfo({
    this.riskAversion,
  });

  factory RiskInfo.fromJson(Map<String, dynamic> json) {
    return RiskInfo(
      riskAversion: json['risk_aversion'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'risk_aversion': riskAversion,
    };
  }

  RiskInfo copyWith({String? riskAversion}) {
    return RiskInfo(
      riskAversion: riskAversion ?? this.riskAversion,
    );
  }
}

class RetirementInfo {
  String? retirementAge;
  String? retirementIncome;

  RetirementInfo({
    this.retirementAge,
    this.retirementIncome,
  });

  factory RetirementInfo.fromJson(Map<String, dynamic> json) {
    return RetirementInfo(
      retirementAge: json['retirement_age'],
      retirementIncome: json['retirement_income'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'retirement_age': retirementAge,
      'retirement_income': retirementIncome,
    };
  }

  RetirementInfo copyWith({
    String? retirementAge,
    String? retirementIncome,
  }) {
    return RetirementInfo(
      retirementAge: retirementAge ?? this.retirementAge,
      retirementIncome: retirementIncome ?? this.retirementIncome,
    );
  }
}

class FamilyMember {
  String? relation;
  String? age;
  String? occupation;
  String? income;

  FamilyMember({
    this.relation,
    this.age,
    this.occupation,
    this.income,
  });

  factory FamilyMember.fromJson(Map<String, dynamic> json) {
    return FamilyMember(
      relation: json['relation'],
      age: json['age'],
      occupation: json['occupation'],
      income: json['income'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'relation': relation,
      'age': age,
      'occupation': occupation,
      'income': income,
    };
  }

  FamilyMember copyWith({
    String? relation,
    String? age,
    String? occupation,
    String? income,
  }) {
    return FamilyMember(
      relation: relation ?? this.relation,
      age: age ?? this.age,
      occupation: occupation ?? this.occupation,
      income: income ?? this.income,
    );
  }
}

class FamilyInfo {
  List<FamilyMember> familyMembers;

  FamilyInfo({
    required this.familyMembers,
  });

  factory FamilyInfo.fromJson(Map<String, dynamic> json) {
    return FamilyInfo(
      familyMembers: (json['family_members'] as List<dynamic>?)
              ?.map((item) => FamilyMember.fromJson(item))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'family_members': familyMembers.map((member) => member.toJson()).toList(),
    };
  }

  FamilyInfo copyWith({List<FamilyMember>? familyMembers}) {
    return FamilyInfo(
      familyMembers: familyMembers ?? List.from(this.familyMembers),
    );
  }
}

class Goal {
  String? goalDetails;

  Goal({
    this.goalDetails,
  });

  factory Goal.fromJson(Map<String, dynamic> json) {
    return Goal(
      goalDetails: json['goal_details'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'goal_details': goalDetails,
    };
  }

  Goal copyWith({String? goalDetails}) {
    return Goal(
      goalDetails: goalDetails ?? this.goalDetails,
    );
  }
}

class GoalInfo {
  List<Goal> goals;

  GoalInfo({
    required this.goals,
  });

  factory GoalInfo.fromJson(Map<String, dynamic> json) {
    return GoalInfo(
      goals: (json['goals'] as List<dynamic>?)
              ?.map((item) => Goal.fromJson(item))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'goals': goals.map((goal) => goal.toJson()).toList(),
    };
  }

  GoalInfo copyWith({List<Goal>? goals}) {
    return GoalInfo(
      goals: goals ?? List.from(this.goals),
    );
  }
}

class UserInfo {
  String? userId;
  BasicInfo basicInfo;
  FinancialInfo financialInfo;
  RiskInfo riskInfo;
  RetirementInfo retirementInfo;
  FamilyInfo familyInfo;
  GoalInfo goalInfo;
  Map<String, dynamic> otherInfo;

  UserInfo({
    this.userId,
    required this.basicInfo,
    required this.financialInfo,
    required this.riskInfo,
    required this.retirementInfo,
    required this.familyInfo,
    required this.goalInfo,
    required this.otherInfo,
  });

  factory UserInfo.empty() {
    return UserInfo(
      basicInfo: BasicInfo(),
      financialInfo: FinancialInfo(),
      riskInfo: RiskInfo(),
      retirementInfo: RetirementInfo(),
      familyInfo: FamilyInfo(familyMembers: []),
      goalInfo: GoalInfo(goals: []),
      otherInfo: {},
    );
  }

  factory UserInfo.fromJson(Map<String, dynamic> json) {
    return UserInfo(
      userId: json['user_id'],
      basicInfo: BasicInfo.fromJson(json['basic_info'] ?? {}),
      financialInfo: FinancialInfo.fromJson(json['financial_info'] ?? {}),
      riskInfo: RiskInfo.fromJson(json['risk_info'] ?? {}),
      retirementInfo: RetirementInfo.fromJson(json['retirement_info'] ?? {}),
      familyInfo: FamilyInfo.fromJson(json['family_info'] ?? {}),
      goalInfo: GoalInfo.fromJson(json['goal_info'] ?? {}),
      otherInfo: Map<String, dynamic>.from(json['other_info'] ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'basic_info': basicInfo.toJson(),
      'financial_info': financialInfo.toJson(),
      'risk_info': riskInfo.toJson(),
      'retirement_info': retirementInfo.toJson(),
      'family_info': familyInfo.toJson(),
      'goal_info': goalInfo.toJson(),
      'other_info': otherInfo,
    };
  }

  UserInfo copyWith({
    String? userId,
    BasicInfo? basicInfo,
    FinancialInfo? financialInfo,
    RiskInfo? riskInfo,
    RetirementInfo? retirementInfo,
    FamilyInfo? familyInfo,
    GoalInfo? goalInfo,
    Map<String, dynamic>? otherInfo,
  }) {
    return UserInfo(
      userId: userId ?? this.userId,
      basicInfo: basicInfo ?? this.basicInfo,
      financialInfo: financialInfo ?? this.financialInfo,
      riskInfo: riskInfo ?? this.riskInfo,
      retirementInfo: retirementInfo ?? this.retirementInfo,
      familyInfo: familyInfo ?? this.familyInfo,
      goalInfo: goalInfo ?? this.goalInfo,
      otherInfo: otherInfo ?? Map.from(this.otherInfo),
    );
  }
}

class UserInfoApiResponse {
  int code;
  String message;
  UserInfo? userInfo;

  UserInfoApiResponse({
    required this.code,
    required this.message,
    this.userInfo,
  });

  factory UserInfoApiResponse.fromJson(Map<String, dynamic> json) {
    return UserInfoApiResponse(
      code: json['code'],
      message: json['message'],
      userInfo: json['user_info'] != null ? UserInfo.fromJson(json['user_info']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'code': code,
      'message': message,
      'user_info': userInfo?.toJson(),
    };
  }
} 